src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -1079,18 +1087,97 @@ public String Replace(char oldChar, char newChar)
             }
         }
 
-        // This method contains the same functionality as StringBuilder Replace. The only difference is that
-        // a new String has to be allocated since Strings are immutable
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        private extern String ReplaceInternal(String oldValue, String newValue);
-
-        public String Replace(String oldValue, String newValue)
+        public string Replace(string oldValue, string newValue)
         {
             if (oldValue == null)
                 throw new ArgumentNullException(nameof(oldValue));
-            // Note that if newValue is null, we treat it like String.Empty.
+            if (oldValue.Length == 0)
+                throw new ArgumentException(SR.Format(SR.Argument_StringZeroLength, nameof(oldValue)));
+
+            // Api behavior: if newValue is null, instances of oldValue are to be removed.
+            if (newValue == null)
+                newValue = string.Empty;
+
+            // Workaround for https://github.com/dotnet/coreclr/issues/16197
+            // Span<int> initialSpan = stackalloc int[StackallocIntBufferSizeLimit];
+            Span<int> initialSpan; StackallocIntBuffer initialBuffer; unsafe { initialSpan = new Span<int>(&initialBuffer, StackallocIntBufferSizeLimit); }
+            var replacementIndices = new ValueListBuilder<int>(initialSpan);
+
+            unsafe
+            {
+                fixed (char* pThis = &_firstChar)
+                {
+                    int matchIdx = 0;
+                    int lastPossibleMatchIdx = this.Length - oldValue.Length;
+                    while (matchIdx <= lastPossibleMatchIdx)
+                    {
+                        char* pMatch = pThis + matchIdx;
+                        for (int probeIdx = 0; probeIdx < oldValue.Length; probeIdx++)
```

Comment:
This is potential candidate for SequenceEquals once it becomes available in CoreLib.